### PR TITLE
GH-276 Apply ComponentCustomizer in other modules

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-cloud/stream-applications/issues/276

Some changes are straightforward as just a `ComponentCustomizer<?>`
injection and its optional usage for the target component customization.
Other changes are more drastic since it is better to use `proxyBeanMethods = false`,
so some direct bean method calls had to be reworked to the bean method argument
injections

* Fix some deprecations from Project Reactor
* Remove `UpdatingMongoDbMessageSource` since `MongoDbMessageSource`
supports now an `update` option in Spring Integration

NOTE: The `ZeroMqSupplierConfiguration` has its own customizer already
like `Consumer<ZMQ.Socket>`, so it can be revised to a more broader solution,
but in the future version.

Some modules just does not make sense to modify for more advance customization
since all their configuration is covered with plain configuration properties.
See `splitter-function` or `wavefront-consumer`

### CODE REVIEW

**Comment by artembilan** [comment]
I believe there is no pressure from end-users, so we may safely consider this change for the next version.
Another thought: we may go baby steps and PR only for individual modules. The review and context of the changes is going to be much easier to track.

Thanks

File: functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java
Comment (onobc) [inline functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java:53]: This is a case where there is already a customizer in Boot (`org.springframework.boot.web.reactive.function.client.WebClientCustomizer`). Should we just respect that as that is what most user's would probably be familiar with?

File: functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java
Comment (onobc) [inline functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java:25]: **[NIT]** copyright ⬆️

File: functions/function/aggregator-function/src/main/java/org/springframework/cloud/fn/aggregator/AggregatorFunctionConfiguration.java
Comment (onobc) [inline functions/function/aggregator-function/src/main/java/org/springframework/cloud/fn/aggregator/AggregatorFunctionConfiguration.java:89]: For consistency should we use an `ObjectProvider` in this particular case since all of the other deps are using that rather than `@Nullable`?

File: functions/consumer/log-consumer/src/main/java/org/springframework/cloud/fn/consumer/log/LogConsumerConfiguration.java
Comment (onobc) [inline functions/consumer/log-consumer/src/main/java/org/springframework/cloud/fn/consumer/log/LogConsumerConfiguration.java:45]: This must be one of the "spring integration API changes" mentioned in the PR description.

File: functions/consumer/rsocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/rsocket/RsocketConsumerConfiguration.java
Comment (onobc) [inline functions/consumer/rsocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/rsocket/RsocketConsumerConfiguration.java:38]: Ahhh yeh, this is a nice simplification of the RSocket API - they dropped the Mono. 👍🏻

File: functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java
Comment (onobc) [inline functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java:64]: Does "beanName" have to be an actual bean name?

File: functions/function/aggregator-function/src/main/java/org/springframework/cloud/fn/aggregator/AggregatorFunctionConfiguration.java
Comment (artembilan) [reply functions/function/aggregator-function/src/main/java/org/springframework/cloud/fn/aggregator/AggregatorFunctionConfiguration.java:89]: I just didn’t know back in days that `@Nullable` will work. Or it just didn’t work in those old Spring versions… 😅

File: functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java
Comment (onobc) [inline functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java:150]: Nice addition!

File: functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java
Comment (artembilan) [reply functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java:25]: Yeah…  In Spring Integration I have Gradle task which updates it for me automatically for all the modified files when I compile project. So, bad habit to forgot 😅.

on the other hand: no claims that this is going to be merged in 2022. So, I’ll just wait for the final decision  😉

File: functions/consumer/rsocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/rsocket/RsocketConsumerConfiguration.java
Comment (artembilan) [reply functions/consumer/rsocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/rsocket/RsocketConsumerConfiguration.java:38]: Well, it still there, but handled internally.

File: functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java
Comment (artembilan) [reply functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java:64]: No. You have better ideas? 🤗

File: functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java
Comment (onobc) [inline functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java:135]: It was not clear to me at first why we do not pass in the `FtpInboundChannelAdapterSpec ftpMessageSource` and then call `ftpMessageSource.get()`. Is it because the framework will have already called `get()` and pass us the resulting message source?

**Comment by sobychacko** [comment]
@artembilan Thank you for this PR! I glanced at the changes, they look good. Once @onobc comments are addressed, I guess we can merge this one.

File: functions/consumer/log-consumer/src/main/java/org/springframework/cloud/fn/consumer/log/LogConsumerConfiguration.java
Comment (artembilan) [reply functions/consumer/log-consumer/src/main/java/org/springframework/cloud/fn/consumer/log/LogConsumerConfiguration.java:45]: Sorry, missed this one. Will add in the next commit. The `log()` now is not a terminal operator in the flow.

File: functions/supplier/mongodb-supplier/src/main/java/org/springframework/cloud/fn/supplier/mongo/UpdatingMongoDbMessageSource.java
Comment (onobc) [inline functions/supplier/mongodb-supplier/src/main/java/org/springframework/cloud/fn/supplier/mongo/UpdatingMongoDbMessageSource.java:45]: Why was this deleted?

File: functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java
Comment (artembilan) [reply functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java:135]: Correct. That’s how I designed SI Java DSL: the `Spec` is a `FactoryBean` by itself, so no need to call it’s result manually and in some cases even dangerous. Perhaps that `get()` has to be hidden somehow, but fully different story…

when we autowire on the other side, we have to deal with the result of the `FactoryBean`. Not sure if Spring describes that somewhere, but I just share our experience what we have so far

File: functions/supplier/mongodb-supplier/src/main/java/org/springframework/cloud/fn/supplier/mongo/UpdatingMongoDbMessageSource.java
Comment (artembilan) [reply functions/supplier/mongodb-supplier/src/main/java/org/springframework/cloud/fn/supplier/mongo/UpdatingMongoDbMessageSource.java:45]: I added that in the commit message

File: functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java
Comment (onobc) [reply functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java:25]: > no claims that this is going to be merged in 2022
Bahaha. Fair enough.

**Comment by artembilan** [comment]
> I guess we can merge this one.

sorry, Soby, not today already: answering from the phone.

File: functions/function/aggregator-function/src/main/java/org/springframework/cloud/fn/aggregator/AggregatorFunctionConfiguration.java
Comment (onobc) [reply functions/function/aggregator-function/src/main/java/org/springframework/cloud/fn/aggregator/AggregatorFunctionConfiguration.java:89]: Makes sense. Should we keep it consistent in this case and move to ObjectProvider to "fit in" w/ the other params?

File: functions/consumer/rsocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/rsocket/RsocketConsumerConfiguration.java
Comment (onobc) [reply functions/consumer/rsocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/rsocket/RsocketConsumerConfiguration.java:38]: Yep, it has to still be there but now the API  avoid the cookie cutter code that everyone did for Mono.cache->flatMap

File: functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java
Comment (onobc) [reply functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java:64]: I noticed that the other specs we call `id` and then that is also what we pass into the customizer. If its simply symbolic name this is fine w/ me.

File: functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java
Comment (onobc) [reply functions/supplier/ftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/ftp/FtpSupplierConfiguration.java:135]: Thx for the explanation.

File: functions/supplier/mongodb-supplier/src/main/java/org/springframework/cloud/fn/supplier/mongo/UpdatingMongoDbMessageSource.java
Comment (onobc) [reply functions/supplier/mongodb-supplier/src/main/java/org/springframework/cloud/fn/supplier/mongo/UpdatingMongoDbMessageSource.java:45]: DOH! Yes you did. My bad. 

> Remove UpdatingMongoDbMessageSource since MongoDbMessageSource
supports now an update option in Spring Integration

I did read the description (but forgot it all by the time  I got all the way to this file). I should have went back to the description before commenting.

File: functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java
Comment (artembilan) [reply functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java:64]: We probably need to look at this from end-user perspective to give them an easy way to make a sense from the names we can provide them. 

I’ll give this more thoughts over night…

**Review by onobc — CHANGES_REQUESTED** [review]
Looks good @artembilan . Thx for taking the time to move to `proxyBeanMethods = false`. I don't mid the extra effort to review it. 

My "changes requested" are minor and up for discussion anyways: 
- nullable vs object provider
- copyright
- weblient customizer

File: functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java
Comment (artembilan) [reply functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java:64]: So, the name we provide to the customizer for end-user consideration must really be a real bean name.
Who knows may be they would like to have an access to that bean eventually...

Or... for consistency with Spring Boot customizer API we just can drop the `beanName` argument altogether...

Comment (onobc) [reply functions/consumer/sftp-consumer/src/main/java/org/springframework/cloud/fn/consumer/sftp/SftpConsumerConfiguration.java:64]: If we do not have a use case yet that would require the bean name, I would be in favor of dropping it altogether (until needed). 

1. makes it consistent w/ the regular "customizer" contract (as you pointed out) 
2. makes it a bit simpler

File: functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java
Comment (onobc) [reply functions/function/http-request-function/src/main/java/org/springframework/cloud/fn/http/request/HttpRequestFunctionConfiguration.java]: Will be handled in #286

**Review by onobc — APPROVED** [review]
@artembilan the only outstanding item is whether or not to yank the bean name from the customizer contract. I approve the PR regardless of your choice on that. Thanks for the updates.

**Review by onobc — APPROVED** [review]

**Pull request merged**
